Firebase/Storage/FIRStorageListTask.m

```diff
@@ -70,7 +70,7 @@ - (void)enqueue {
     // listAll() doesn't set a pageSize as this allows Firebase Storage to determine how many items
     // to return per page. This removes the need to backfill results if Firebase Storage filters
     // objects that are considered invalid (such as items with two consecutive slashes).
-    if (strongSelf->_pageSize) {
+    if (strongSelf->_pageSize != nil) {
```

Comment:
I prefer thins type of check on ObjC objects just from the readability perspective. I wonder, when it behaves differently?

Reply:
The only thing I could find was [this](https://www.mikeash.com/pyblog/friday-qa-2012-12-14-objective-c-pitfalls.html), under the section "Cast to BOOL"

Reply:
Interesting... I didn't know this is possible with objects... Also, it means that the `init` Xcode snippet is buggy because it generates `if (self) { ...`. If it is as bad as described (6% chance) then we really need to update it through our codebase.

Reply:
Or it happens only when you explicitly cast with `(BOOL)`?

Reply:
Looks like it's [incidentally correct on 64-bit iOS](https://stackoverflow.com/questions/31267325/bool-with-64-bit-on-ios/31267502), but not 32-bit iOS and probably not other platforms. It is something we should fix throughout our codebase.

Reply:
Well, I think, we are OK here. The described issue is for the Objective-C `BOOL` only. When we use a pointer in conditions it is not converted to the Objective-C `BOOL`, but the C casting rules are used, so the pointer will be casted to an `int` type. In this case we should not expect any issue.
So basically:
Safe option: `if (self) { ... ` 
Unsafe option:: `if ((BOOL)self) { ... `

I hope I am not missing anything :)

Reply:
Explicitly casting to `BOOL` is the problem. General boolean evaluation is not subject to this problem:

```c
#import <Foundation/Foundation.h>

#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>

int main() {
  for (uintptr_t value = 0; value <= 256; value += 8) {
    void* pointer = (void*)value;
    BOOL coercedBool = (BOOL)pointer;

    const char* matched = "false";
    if (pointer) {
      matched = "true";
    }

    printf("BOOL %" PRIuPTR " = %d, if: %s\n", value, coercedBool, matched);
  }
}
```
Put it in a file `test.m` and run with:
```
clang test.m && ./a.out
```

Output:
```
BOOL 0 = 0, if: false
BOOL 8 = 8, if: true
BOOL 16 = 16, if: true
BOOL 24 = 24, if: true
BOOL 32 = 32, if: true
BOOL 40 = 40, if: true
BOOL 48 = 48, if: true
BOOL 56 = 56, if: true
BOOL 64 = 64, if: true
BOOL 72 = 72, if: true
BOOL 80 = 80, if: true
BOOL 88 = 88, if: true
BOOL 96 = 96, if: true
BOOL 104 = 104, if: true
BOOL 112 = 112, if: true
BOOL 120 = 120, if: true
BOOL 128 = -128, if: true
BOOL 136 = -120, if: true
BOOL 144 = -112, if: true
BOOL 152 = -104, if: true
BOOL 160 = -96, if: true
BOOL 168 = -88, if: true
BOOL 176 = -80, if: true
BOOL 184 = -72, if: true
BOOL 192 = -64, if: true
BOOL 200 = -56, if: true
BOOL 208 = -48, if: true
BOOL 216 = -40, if: true
BOOL 224 = -32, if: true
BOOL 232 = -24, if: true
BOOL 240 = -16, if: true
BOOL 248 = -8, if: true
BOOL 256 = 0, if: true
```
What this means is that as the pointer is an even multiple of 256, casting to `BOOL` will wrap around. Checking if a pointer is non-nil in a boolean context does not cast to `BOOL` and so is not subject to the same problem.

That is, `if (self) { ... }` and `if (self != nil) { ... }` are equivalently reliable.

This warning you got comes about because you're testing an `NSNumber*` object pointer for whether or not it's set, but the value of a set `NSNumber` could be false. It's a common enough error to accidentally check object-is-set when you really meant object-has-a-non-false-value that the old usage is flagged.

Reply:
Thanks Gil :)